File: lib/domain/insights/caffeine_drink_profile.dart
import 'dart:math' as math;

import 'package:freezed_annotation/freezed_annotation.dart';

import '../model/caffeine_models.dart';
import '../preferences/body_profile.dart';
import '../preferences/caffeine_preferences.dart';
import 'caffeine_insight_calculator.dart';

part 'caffeine_drink_profile.freezed.dart';

/// One drink, on its own: what it did, what it is still doing, and when it will be done.
///
/// Nothing here is new arithmetic. [CaffeineInsightCalculator.contributionMg] already
/// answers "how much of THIS drink is still in you at that moment" — it has to, because
/// the whole-day curve is the sum of exactly these. All this does is ask it about one
/// drink, repeatedly, which is the difference between a number in a total and a thing you
/// can understand.
///
/// The question a drinker actually has is "will this coffee still be in me at bedtime",
/// so that is the question this answers: the peak and when it fell, what is left right
/// now, when half of it will be gone, and when it will effectively be gone.

/// Below this, a drink is finished. Not zero: the model decays exponentially and never
/// reaches zero, so a "gone" that waited for zero would never come.
const double caffeineNegligibleMg = 5.0;

/// How far past a drink the profile looks before giving up on it fading.
///
/// Thirty-six hours, not twenty-four. A large dose really does take longer than a day to
/// fall below [caffeineNegligibleMg] -- a 200mg energy drink is still carrying more than
/// that a full day later -- and a 24-hour horizon simply reported "we do not know when
/// this goes away" for exactly the drinks whose staying power is most worth knowing.
const Duration caffeineProfileHorizon = Duration(hours: 36);

/// How finely the drink's curve is sampled.
const Duration caffeineProfileStep = Duration(minutes: 10);

@freezed
abstract class CaffeineDrinkProfile with _$CaffeineDrinkProfile {
  const CaffeineDrinkProfile._();

  const factory CaffeineDrinkProfile({
    required CaffeineEntry entry,

    /// This drink's own rise and fall — not the day's.
    required List<CaffeinePoint> curve,

    /// The most of this drink that was ever in the body at once, and when.
    ///
    /// Lower than the dose, always: absorption takes time, and elimination has begun
    /// before absorption has finished. A 95mg coffee never puts 95mg in you at once.
    required double peakMg,
    required DateTime peakTime,

    /// What is left of it now. Zero before it was drunk.
    required double currentMg,

    /// When half of the peak has gone, and when what remains stops mattering.
    /// Null when the drink has not faded within [caffeineProfileHorizon].
    required DateTime? halfGoneTime,
    required DateTime? goneTime,
  }) = _CaffeineDrinkProfile;

  /// Whether this drink is still doing anything worth speaking of.
  bool get isActive => currentMg >= caffeineNegligibleMg;
}

/// Works [entry] out on its own, against the same model the whole-day curve uses — so the
/// number a drink shows here and the bump it makes in the day's curve are the same number,
/// and can never disagree.
CaffeineDrinkProfile caffeineDrinkProfile({
  required CaffeineEntry entry,
  required DateTime now,
  required CaffeinePreferences preferences,
  BodyProfile bodyProfile = const BodyProfile(),
}) {
  final start = entry.startTime;
  final end = start.add(caffeineProfileHorizon);

  final curve = <CaffeinePoint>[];
  var peakMg = 0.0;
  var peakTime = start;

  for (var time = start;
      !time.isAfter(end);
      time = time.add(caffeineProfileStep)) {
    final value = CaffeineInsightCalculator.contributionMg(
      entry: entry,
      at: time,
      preferences: preferences,
      bodyProfile: bodyProfile,
    );
    curve.add(CaffeinePoint(time: time, valueMg: value));
    if (value > peakMg) {
      peakMg = value;
      peakTime = time;
    }
  }

  final currentMg = CaffeineInsightCalculator.contributionMg(
    entry: entry,
    at: now,
    preferences: preferences,
    bodyProfile: bodyProfile,
  );

  return CaffeineDrinkProfile(
    entry: entry,
    curve: curve,
    peakMg: peakMg,
    peakTime: peakTime,
    currentMg: currentMg,
    // Both are read off the curve AFTER the peak. Before it the drink is still being
    // absorbed and is on its way up — a threshold crossed on the way up is not the drink
    // fading, it is the drink arriving.
    halfGoneTime: _fallsBelow(curve, peakTime, peakMg / 2.0),
    goneTime: _fallsBelow(curve, peakTime, caffeineNegligibleMg),
  );
}

/// The first moment after [afterTime] that the curve drops below [threshold], or null if
/// it never does within the horizon.
DateTime? _fallsBelow(
  List<CaffeinePoint> curve,
  DateTime afterTime,
  double threshold,
) {
  if (threshold <= 0) return null;
  for (final point in curve) {
    if (point.time.isBefore(afterTime)) continue;
    if (point.valueMg < threshold) return point.time;
  }
  return null;
}

/// The highest point of any of [profiles] — what a set of per-drink charts should share as
/// their y axis, so a small drink next to a large one LOOKS small.
double caffeineProfilePeak(Iterable<CaffeineDrinkProfile> profiles) =>
    profiles.fold<double>(0.0, (peak, profile) => math.max(peak, profile.peakMg));

File: app/src/main/kotlin/tech/mmarca/openvitals/domain/insights/CaffeineDrinkProfile.kt
package tech.mmarca.openvitals.domain.insights

import java.time.Duration
import java.time.Instant
import tech.mmarca.openvitals.domain.model.CaffeineEntry
import tech.mmarca.openvitals.domain.model.CaffeinePoint
import tech.mmarca.openvitals.domain.preferences.BodyProfile
import tech.mmarca.openvitals.domain.preferences.CaffeinePreferences

/**
 * One drink on its own: its peak, what is left now, when half is gone and
 * when it is effectively gone. The same arithmetic as the day curve,
 * asked about one drink.
 */
data class CaffeineDrinkProfile(
    val entry: CaffeineEntry,

    /** This drink's own rise and fall — not the day's. */
    val curve: List<CaffeinePoint>,

    /** The most of this drink ever in the body at once. Lower than the dose: elimination starts early. */
    val peakMg: Double,
    val peakTime: Instant,

    /** What is left of it now. Zero before it was drunk. */
    val currentMg: Double,

    /** When half the peak has gone. Null when the drink has not faded within [CaffeineProfileHorizon]. */
    val halfGoneTime: Instant?,
    val goneTime: Instant?,
) {
    /** Whether this drink is still doing anything worth speaking of. */
    val isActive: Boolean get() = currentMg >= CaffeineNegligibleMg
}

/** Below this a drink is finished. Not zero: the decay never reaches zero. */
const val CaffeineNegligibleMg: Double = 5.0

/** How far past a drink the profile looks. 36 hours: a 200mg drink is still over the floor a day later. */
val CaffeineProfileHorizon: Duration = Duration.ofHours(36)

/** How finely the drink's curve is sampled. */
val CaffeineProfileStep: Duration = Duration.ofMinutes(10)

/**
 * Works [entry] out against the same model the day curve uses, so the two
 * can never disagree. Preferences are normalized for the same reason.
 */
fun caffeineDrinkProfile(
    entry: CaffeineEntry,
    now: Instant,
    preferences: CaffeinePreferences,
    bodyProfile: BodyProfile = BodyProfile(),
): CaffeineDrinkProfile {
    val normalized = preferences.normalized()
    val start = entry.startTime
    val end = start.plus(CaffeineProfileHorizon)

    val curve = mutableListOf<CaffeinePoint>()
    var peakMg = 0.0
    var peakTime = start

    var time = start
    while (!time.isAfter(end)) {
        val value = CaffeineInsightCalculator.contributionMg(entry, time, normalized, bodyProfile)
        curve += CaffeinePoint(time, value)
        if (value > peakMg) {
            peakMg = value
            peakTime = time
        }
        time = time.plus(CaffeineProfileStep)
    }

    return CaffeineDrinkProfile(
        entry = entry,
        curve = curve,
        peakMg = peakMg,
        peakTime = peakTime,
        currentMg = CaffeineInsightCalculator.contributionMg(entry, now, normalized, bodyProfile),
        // Read after the peak: a threshold crossed on the way up is the drink arriving.
        halfGoneTime = curve.fallsBelow(peakTime, peakMg / 2.0),
        goneTime = curve.fallsBelow(peakTime, CaffeineNegligibleMg),
    )
}

/** The highest point of any of [profiles], for a shared y axis. */
fun caffeineProfilePeak(profiles: Iterable<CaffeineDrinkProfile>): Double =
    profiles.fold(0.0) { peak, profile -> maxOf(peak, profile.peakMg) }

/** The first moment after [afterTime] the curve drops below [threshold], or null. */
private fun List<CaffeinePoint>.fallsBelow(afterTime: Instant, threshold: Double): Instant? {
    if (threshold <= 0.0) return null
    for (point in this) {
        if (point.time.isBefore(afterTime)) continue
        if (point.valueMg < threshold) return point.time
    }
    return null
}
